src/client.rs

```diff
@@ -74,6 +77,165 @@ pub extern "C" fn rustls_client_config_builder_build(
     }
 }
 
+/// Currently just a placeholder with no accessors yet.
+/// https://docs.rs/rustls/0.19.0/rustls/struct.RootCertStore.html
+#[allow(non_camel_case_types)]
+pub struct rustls_root_cert_store {
+    _private: [u8; 0],
+}
+
+/// Input to a custom certificate verifier callback. See
+/// rustls_client_config_builder_dangerous_set_certificate_verifier().
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct rustls_verify_server_cert_params<'a> {
+    end_entity: rustls_slice_bytes<'a>,
+    intermediates: rustls_slice_slice_bytes<'a>,
+    roots: *const rustls_root_cert_store,
+    dns_name: rustls_str<'a>,
+    ocsp_response: rustls_slice_bytes<'a>,
+}
+
+/// User-provided input to a custom certificate verifier callback. See
+/// rustls_client_config_builder_dangerous_set_certificate_verifier().
+#[allow(non_camel_case_types)]
+type rustls_verify_server_cert_user_data = *mut libc::c_void;
+
+// According to the nomicon https://doc.rust-lang.org/nomicon/ffi.html#the-nullable-pointer-optimization):
+// > Option<extern "C" fn(c_int) -> c_int> is a correct way to represent a
+// > nullable function pointer using the C ABI (corresponding to the C type int (*)(int)).
+// So we use Option<...> here. This is the type that is passed from C code.
+#[allow(non_camel_case_types)]
+type rustls_verify_server_cert_callback = Option<
+    unsafe extern "C" fn(
+        userdata: rustls_verify_server_cert_user_data,
+        params: *const rustls_verify_server_cert_params,
+    ) -> rustls_result,
+>;
+
+// This is the same as a rustls_verify_server_cert_callback after unwrapping
+// the Option (which is equivalent to checking for null).
+type VerifyCallback = unsafe extern "C" fn(
+    userdata: rustls_verify_server_cert_user_data,
+    params: *const rustls_verify_server_cert_params,
+) -> rustls_result;
+
+// An implementation of rustls::ServerCertVerifier based on a C callback.
+struct Verifier {
+    callback: VerifyCallback,
+    userdata: rustls_verify_server_cert_user_data,
+}
+
+/// Safety: Verifier is Send because we don't allocate or deallocate any of its
+/// fields.
+unsafe impl Send for Verifier {}
+/// Safety: Verifier is Sync if the C code that passes us a callback that
+/// obeys the concurrency safety requirements documented in
+/// rustls_client_config_builder_dangerous_set_certificate_verifier.
+unsafe impl Sync for Verifier {}
+
+impl rustls::ServerCertVerifier for Verifier {
+    fn verify_server_cert(
+        &self,
+        roots: &RootCertStore,
+        presented_certs: &[Certificate],
+        dns_name: DNSNameRef<'_>,
+        ocsp_response: &[u8],
+    ) -> Result<ServerCertVerified, TLSError> {
+        let cb = self.callback;
+        let dns_name: &str = dns_name.into();
+        let dns_name: rustls_str = match dns_name.try_into() {
+            Ok(r) => r,
+            Err(_) => return Err(TLSError::General("NUL byte in SNI".to_string())),
```

Comment:
nit: I know this is the only possible failure case from `rustls_str::from` but it'd be nice if that `From` impl's associated error type was `String` and it returned "NUL byte in SNI" on failure so that you wouldn't have to assume that was the cause of the error here in this method.

Reply:
I updated the `Error` associated type for `TryFrom` to be a type alias `NulByte`, as a way of documenting that's the only error that will come out. I wasn't able to find a way to make that clear at this match size. `Err(())` works but `Err(NulByte)` doesn't.